### ISSUE TITLE
Build error not showing in hsdev branch

### ISSUE THREAD

**Issue by tuvistavie** [SublimeHaskell/SublimeHaskell#229]
Hi,
I am having some issues when trying to use the hsdev branch.

The main issue is that when the build fails, nothing shows up, and in the console I get the following message

```
Traceback (most recent call last):
  File "./threading.py", line 901, in _bootstrap_inner
  File "./threading.py", line 858, in run
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 154, in wait_for_chain_to_complete
    parse_output_messages_and_show(view, msg, cabal_project_dir, exit_code, errmsg)
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 192, in parse_output_messages_and_show
    set_global_error_messages(parsed_messages)
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 112, in set_global_error_messages
    ERRORS[m.filename][m.start.line].append(m)
AttributeError: 'str' object has no attribute 'line'
```

Another issue, which is I think related, is that hsdev works when I start it manually, but 
I get errors when Sublime starts it:

```
Sublime Haskell: hsdev scan --project /home/daniel/Documents/sample-project
Sublime Haskell: hsdev returns error: Exception, details: {'what': 'src: canonicalizePath: does not exist (No such file or directory)'}
```

Here are info about my environment:

```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.10.1
$ hsdev version
0.1.4.0
$ cabal --version
cabal-install version 1.22.6.0
using version 1.22.2.0 of the Cabal library
```

The project is in a cabal sandbox.

The build (and only the build, not hsdev as the version is I think not compatible) seems to work properly when I try the hsdev-socket branch.
Everything works fine when I use the master branch.

Thanks for you help.

**Comment by mvoidex** [comment]
Can you try to run these `hsdev` commands from console (with Sublime closed):

    hsdev start
    hsdev scan --project /home/daniel/Documents/sample-project
    hsdev stop

Will it return same error?

**Comment by tuvistavie** [comment]
Thanks for the quick reply.

I tried to run it with Sublime closed, if I run `hsdev start` inside the project everything goes fine, however if I run `hsdev start` from some other location, I get the same error.
Here is the full output of `scan` with the paths of the project I am actually working on.

```
{"notify":{"status":"working","child":null,"progress":null,"params":{"project":"/home/daniel/Documents/organizations/claudetech/applify/imageserver","name":"imageserver","type":"project"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{"name":"/home/daniel/Documents/organizations/claudetech/applify/imageserver","type":"path","file":"/home/daniel/Documents/organizations/claudetech/applify/imageserver"},"task":"scanning"},"progress":null,"params":{"project":"/home/daniel/Documents/organizations/claudetech/applify/imageserver","name":"imageserver","type":"project"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{"name":"/home/daniel/Documents/organizations/claudetech/applify/imageserver","type":"path","file":"/home/daniel/Documents/organizations/claudetech/applify/imageserver"},"task":"scanning"},"progress":null,"params":{"project":"/home/daniel/Documents/organizations/claudetech/applify/imageserver","name":"imageserver","type":"project"},"task":"scanning"}}
{"error":"Exception","details":{"what":"src: canonicalizePath: does not exist (No such file or directory)"}}
```

**Comment by mvoidex** [comment]
I think I found bug. If you want, you can try updated version of [hsdev](https://github.com/mvoidex/hsdev) from GitHub, while i'm installing Ubuntu to check if it's fixed.

**Comment by tuvistavie** [comment]
Hi,

The scan seems to be fixed, but the build is still in the same situation and other functionalities, such as `ghc-mod type` are not working.

Here is a log (with "log": 5 in settings) of the startup process, and what I get after trying to show type.

```
Sublime Haskell: Starting hsdev server command: ['hsdev', 'start', '--port', '4567', '--cache', '/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/./hsdev', '--log', '/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/./hsdev/hsdev.log']
Sublime Haskell: Starting hsdev server
Sublime Haskell: connecting to hsdev server...
Sublime Haskell: connecting to hsdev server...
Sublime Haskell: failed to connect to hsdev serverSublime Haskell: failed to connect to hsdev server

Sublime Haskell: connecting to hsdev server...
Sublime Haskell: connecting to hsdev server...
Sublime Haskell: failed to connect to hsdev server
Sublime Haskell: failed to connect to hsdev server
warning: mnemonic b not found in menu caption SFTP/FTP
warning: mnemonic r not found in menu caption SublimeREPL
Emmet: No need to update PyV8
Sublime Haskell: connecting to hsdev server...
Sublime Haskell: connecting to hsdev server...
Sublime Haskell: connected to hsdev server
Sublime Haskell: connected to hsdev server
Sublime Haskell: hsdev link
Sublime Haskell: hsdev scan --cabal
Package Control: Skipping automatic upgrade, last run at 2015-07-21 14:20:23, next run at 2015-07-21 15:20:23 or after
Sublime Haskell: hsdev module --file /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs
Sublime Haskell: hsdev ghc-mod type 22 5 --file /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs --ghc -i /home/daniel/Documents/organizations/claudetech/applify/imageserver
Sublime Haskell: hsdev returns error: Module not found, details: {}
```

Then, if I run

```
hsdev scan --project /home/daniel/Documents/organizations/claudetech/applify/imageserver
```

manually and try to show the type again, I get

```
Sublime Haskell: hsdev module --file /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs
Sublime Haskell: hsdev ghc-mod type 22 5 --file /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs --ghc -i /home/daniel/Documents/organizations/claudetech/applify/imageserver
Sublime Haskell: completions: 0
```

So I suppose the file has been found, but for some reason the type cannot be inferred.
The cursor is pointing at `putStrLn` so it is probably not a problem of libraries not found or anything.

When I save (with auto build activated), if the projects has no error everything goes fine:

```
Sublime Haskell: hsdev projects
Sublime Haskell: running build commands: [['cabal', 'build'], ['cabal', 'build', '-v0', '--ghc-options=-fforce-recomp -Wall -fno-code']]
Sublime Haskell: total time to mark 0 diagnostics: 0.00048099999999973164 seconds
```

However, if I have an error, obviously the build fails, but the error is not shown and the console show the error I wrote in my first post

```
Sublime Haskell: running build commands: [['cabal', 'build'], ['cabal', 'build', '-v0', '--ghc-options=-fforce-recomp -Wall -fno-code']]
Sublime Haskell: completions: 0
Exception in thread Thread-545:
Traceback (most recent call last):
  File "./threading.py", line 901, in _bootstrap_inner
  File "./threading.py", line 858, in run
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 154, in wait_for_chain_to_complete
    parse_output_messages_and_show(view, msg, cabal_project_dir, exit_code, errmsg)
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 192, in parse_output_messages_and_show
    set_global_error_messages(parsed_messages)
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 112, in set_global_error_messages
    ERRORS[m.filename][m.start.line].append(m)
AttributeError: 'str' object has no attribute 'line'
```

After trying to debug a little myself, it seems that `m.start` is a string with the value of the line number containing the build error, but I was unable to find out why.

Thank you for your help. Please tell me if you need any other information.

**Comment by mvoidex** [comment]
I've made some fixes that worked for me. Can you try?

**Comment by tuvistavie** [comment]
Hi, thanks for the help.

For the build, it seems to be getting better:

I tried to build with an error in the project and got the following error message:

```
Exception in thread Thread-59:
Traceback (most recent call last):
  File "./threading.py", line 901, in _bootstrap_inner
  File "./threading.py", line 858, in run
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 161, in wait_for_chain_to_complete
    parse_output_messages_and_show(view, msg, cabal_project_dir, exit_code, errmsg)
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 194, in parse_output_messages_and_show
    parsed_messages = parse_output_messages(view, base_dir, stderr)
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 394, in parse_output_messages
    return list(map(to_error, matches))
  File "/home/daniel/.config/sublime-text-3/Packages/SublimeHaskell/parseoutput.py", line 384, in to_error
    column = type_column_to_sublime_column(view, line, column)
NameError: global name 'type_column_to_sublime_column' is not defined
```

after looking around `parseoutput.py` and doing the following change, I managed to get the panel output error back with the proper error.

```diff
diff --git a/parseoutput.py b/parseoutput.py
index e85e20a..59d71e8 100644
--- a/parseoutput.py
+++ b/parseoutput.py
@@ -380,8 +380,9 @@ def parse_output_messages(view, base_dir, text):
 
     def to_error(m):
         filename, line, column, messy_details = m.groups()
+        line, column = int(line), int(column)
 
-        column = type_column_to_sublime_column(view, line, column)
+        column = ghc_column_to_sublime_column(view, line, column)
         line = line - 1
         return OutputMessage(
             # Record the absolute, normalized path.
```

However, the line with the problem does not get highlighted so maybe something else needs to be fixed.

For the hsdev part, I found out that `hsdev scan --cabal` is not scanning properly: here is the output

http://pastebin.com/RTETm4mM

and the log with `--project`

http://pastebin.com/7T6KTHTz

Here is my .cabal file if it can help: https://gist.github.com/6c07ca94883758cf70e0

After running `scan` with the `--project` option, I get completions working again (`Sublime Haskell: completions: 278` appears in the console), but show type never returns any result:

```
$ hsdev ghc-mod type 19 6 --file /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs
[]
```

Let me know if you need any other info.

**Comment by mvoidex** [comment]
Thanks.

> For the hsdev part, I found out that  hsdev scan --cabal  is not scanning properly: here is the output

Everything seems to be ok, `hsdev scan` scans all modules at once.

> but show type never returns any result

I think I've reproduced this bug. Try to run these commands from `/home/daniel/Documents/organizations/claudetech/applify/imageserver`

    ghc-mod type src/Main.hs 19 6
    ghc-mod -g "-i src" type src/Main.hs 19 6
    hsdev ghc-mod type 19 6 --file src/Main.hs
    hsdev ghc-mod type 19 6 --file src/Main.hs -g "-i src"

**Comment by tuvistavie** [comment]
Here is the output for the commands, it seems to work when passing `-g "-i src"`.

```
$ ghc-mod type src/Main.hs 19 6            
Warning: Could not find a component assignment, falling back to guessed GHC
         options.
EXCEPTION: types:
           Could not find module ‘Storage.Backends.Local’
Use -v to see a list of the files searched for.
Could not find module ‘DB.Backends.Redis’
Use -v to see a list of the files searched for.
Could not find module ‘App.Handlers’
Use -v to see a list of the files searched for.
$ ghc-mod -g "-i src" type src/Main.hs 19 6       
Warning: Could not find a component assignment, falling back to guessed GHC
         options.
19 5 19 15 "FilePath"
18 8 23 37 "IO ()"
18 1 23 37 "IO ()"
$ hsdev ghc-mod type 19 6 --file src/Main.hs            
[]
$ hsdev ghc-mod type 19 6 --file src/Main.hs -g "-i src"
[{"expr":"currentDir","region":{"to":{"line":19,"column":15},"from":{"line":19,"column":5}},"type":"FilePath"},{"expr":"do\n    currentDir <- getCurrentDirectory\n    let storage = L.makeStorage $ joinPath [currentDir, \"uploads\"]\n    connection <- R.connect R.defaultConnectInfo\n    let db = DR.makeDB connection\n    scotty 4000 $ handler storage db","region":{"to":{"line":23,"column":37},"from":{"line":18,"column":8}},"type":"IO ()"},{"expr":"main = do\n    currentDir <- getCurrentDirectory\n    let storage = L.makeStorage $ joinPath [currentDir, \"uploads\"]\n    connection <- R.connect R.defaultConnectInfo\n    let db = DR.makeDB connection\n    scotty 4000 $ handler storage db","region":{"to":{"line":23,"column":37},"from":{"line":18,"column":1}},"type":"IO ()"}]
```

By the way, it works only after I run `hsdev scan --project /path/to/project`:

```sh
$ hsdev start
Server started at port 4567
$ hsdev scan --cabal                                                                       
{"notify":{"status":"working","child":null,"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"getting list of cabal modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"getting list of cabal modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"getting list of changed modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"getting list of changed modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{"modules":[]},"task":"removing obsolete modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{"modules":[]},"task":"removing obsolete modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"loading modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"loading modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"loading docs"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"loading docs"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"ok","child":null,"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
[]
$ hsdev ghc-mod type 19 6 --file src/Main.hs -g "-i src"
{"error":"File '/home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs' not found","details":{}}
$ hsdev scan --project /home/daniel/Documents/organizations/claudetech/applify/imageserver
.....the long output I pasted on pastebin.....
$ hsdev ghc-mod type 19 6 --file src/Main.hs -g "-i src"
[{"expr":"currentDir","region":{"to":{"line":19,"column":15},"from":{"line":19,"column":5}},"type":"FilePath"},{"expr":"do\n    currentDir <- getCurrentDirectory\n    let storage = L.makeStorage $ joinPath [currentDir, \"uploads\"]\n    connection <- R.connect R.defaultConnectInfo\n    let db = DR.makeDB connection\n    scotty 4000 $ handler storage db","region":{"to":{"line":23,"column":37},"from":{"line":18,"column":8}},"type":"IO ()"},{"expr":"main = do\n    currentDir <- getCurrentDirectory\n    let storage = L.makeStorage $ joinPath [currentDir, \"uploads\"]\n    connection <- R.connect R.defaultConnectInfo\n    let db = DR.makeDB connection\n    scotty 4000 $ handler storage db","region":{"to":{"line":23,"column":37},"from":{"line":18,"column":1}},"type":"IO ()"}]
```

**Comment by mvoidex** [comment]
Thank for information. I'll try to fix it this week

**Comment by tomv564** [comment]
This appears to have fixed the build-on-save functionality, thanks!
The following fixes highlighting the lines in the view:

```
 File "/Users/tomv/Library/Application Support/Sublime Text 3/Packages/SublimeHaskell/parseoutput.py", line 101, in to_region_in_view
    return trim_region(view.line(self.start.to_point_of_view()))
TypeError: to_point_of_view() missing 1 required positional argument: 'view'

``` 

This line should probably be: 

``` python
return trim_region(view, view.line(self.start.to_point_of_view(view)))
```

**Comment by tuvistavie** [comment]
The above change worked for me too, thanks.

I think I also found the issue for `show_type`.
In [this commit](https://github.com/mvoidex/hsdev/commit/1c83a56fca2691554142fd3fadca6b2834d14bdb) the usage of `hsinspect` seems to have changed, however `sublime_haskell_common.get_source_dir` still uses the old syntax with the `cabal` argument,
making it impossible to parse the `cabal` file, which is probably why the value
of the `-i` option passed to ghc-mod is wrong.

```diff
diff --git a/sublime_haskell_common.py b/sublime_haskell_common.py
index 788ec4c..5178cc0 100644
--- a/sublime_haskell_common.py
+++ b/sublime_haskell_common.py
@@ -484,7 +484,7 @@ def get_source_dir(filename):
         return os.path.dirname(filename)
 
     _project_name, cabal_file = get_cabal_in_dir(cabal_dir)
-    exit_code, out, err = call_and_wait(['hsinspect', 'cabal', cabal_file])
+    exit_code, out, err = call_and_wait(['hsinspect', cabal_file])
 
     if exit_code == 0:
         info = json.loads(out)

```

restored the show type and insert type functionalities for me.
I just sent a PR with the changes that @tomv564 mentionned, as well as the above change.

The only remaining issue (I know about) now is that `scan --cabal` does not seem to detect any file, and I therefore need to run `hsdev scan --project` manually in order to get hsdev functionalities working.

**Comment by mvoidex** [comment]
Fix for `hsinspect` issue https://github.com/SublimeHaskell/SublimeHaskell/commit/fcb7ddc378f1ac150aa7dcf4fec78db12c3e6d96

@tuvistavie,

> I therefore need to run  hsdev scan --project  manually in order to get hsdev functionalities working

Can you describe more detail? In my case all projects found in project folders are scanned automatically. When I open some .hs file, which is project-related, its project is scanned too.

**Comment by tuvistavie** [comment]
Thanks for the help.

Here is Sublime Haskell log on startup:

```
Sublime Haskell: connecting to hsdev server...
Sublime Haskell: connected to hsdev server
Sublime Haskell: hsdev link
Sublime Haskell: hsdev scan --cabal
Sublime Haskell: drop prepared completions
Sublime Haskell: preparing completions for /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs
Sublime Haskell: hsdev module --file /home/daniel/Documents/organizations/claudetech/applify/imageserver/src/Main.hs
Sublime Haskell: hsdev returns error: Module not found, details: {}
Sublime Haskell: completions: 0
Sublime Haskell: completions: 0
```

Here is the full output of `hsdev scan --cabal` when I run it from the root of my project (where the .cabal file lives)

```
{"notify":{"status":"working","child":null,"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"getting list of cabal modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"getting list of cabal modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"getting list of changed modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"getting list of changed modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{"modules":[]},"task":"removing obsolete modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{"modules":[]},"task":"removing obsolete modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"loading modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"loading modules"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":null,"params":{},"task":"loading docs"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"working","child":{"status":"ok","child":null,"progress":null,"params":{},"task":"loading docs"},"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
{"notify":{"status":"ok","child":null,"progress":null,"params":{"sandbox":"cabal","name":"cabal","type":"cabal"},"task":"scanning"}}
[]
```

When I run `hsdev scan --project /path/to/project`, I have a lot of outputs of this kind:

```
{"notify":{"status":"working","child":{"status":"working","child":null,"progress":{"current":2,"total":26},"params":{"name":"/home/daniel/Documents/organizations/claudetech/applify/imageserver/src/DB/Backends.hs","module":{"project":"/home/daniel/Documents/organizations/claudetech/applify/imageserver/image-server.cabal","file":"/home/daniel/Documents/organizations/claudetech/applify/imageserver/src/DB/Backends.hs"},"type":"module"},"task":"scanning"},"progress":null,"params":{"project":"/home/daniel/Documents/organizations/claudetech/applify/imageserver","name":"imageserver","type":"project"},"task":"scanning"}}
```

which is where the files apparently get scanned.
As posted above, the full output is available [on pastebin](http://pastebin.com/7T6KTHTz) and my `cabal` file is available at https://gist.github.com/6c07ca94883758cf70e0

Let me know if you need any other info.

**Comment by mvoidex** [comment]
I'll close issue, because were many changes. Let me know, if something isn't working for you
